lib/perl/Genome/InstrumentData/AlignmentResult/Merged.pm

```diff
@@ -219,24 +226,70 @@ sub create {
         for my $alignment (@alignments) {
             $alignment->add_user(user => $self, label => 'uses');
         }
-
-        return 1;
-    };
-    if (my $error = $@) {
-        $tx->rollback();
-        die $error;
-    } elsif ($rv ne 1) {
+    }
+    catch {
         $tx->rollback();
-        $self->error_message('Unexpected return value: ' . $rv);
-        die $self->error_message;
-    } else {
-        $tx->commit();
+        die $self->error_message('Merge failed due to error: ' . $_);
     }
+    finally {
+        for my $allocation (@temp_allocations) {
+            $allocation->delete;
+        }
+        unless ($_) {
+            $tx->commit();
+            $self->debug_message('Merge done');
+        }
+    };
 
     $self->debug_message("Resizing the disk allocation...");
     if ($self->_disk_allocation) {
-        unless ($self->_disk_allocation->reallocate) {
-            $self->warning_message("Failed to reallocate disk allocation: " . $self->_disk_allocation->id);
+        unless (eval { $self->_disk_allocation->reallocate }) {
+            $self->warning_message("Failed to reallocate my disk allocation with id (%s). Eval returned (%s) ", $self->_disk_allocation->id, $@);
+        }
+        $self->output_dir($self->_disk_allocation->absolute_path); #update if was moved
+    }
+
+    #purge per lane alignment along with its .bai and md5 files, but
+    #create header files and keep flagstat files for the future use
+    for my $alignment ($self->collect_individual_alignments) {
+        for my $bam_path ($alignment->alignment_bam_file_paths) {
+            my $header = $bam_path . '.header';
+            unless (-s $header) {
+                my $sam_path = Genome::Model::Tools::Sam->path_for_samtools_version($self->samtools_version);
+                my $cmd = "$sam_path view -H $bam_path > $header";
+                Genome::Sys->shellcmd(
+                    cmd => $cmd,
+                    input_files  => [$bam_path],
+                    output_files => [$header],
+                );
+            }
+            my $flagstat = $bam_path . '.flagstat';
+            unless (-s $flagstat) {
+                my $cmd = Genome::Model::Tools::Sam::Flagstat->create(
+                    bam_file    => $bam_path,
+                    output_file => $flagstat,
+                    use_version => $self->samtools_version,
+                );
+
+                unless ($cmd->execute) {
+                    die $self->error_message("Fail to run flagstat on $bam_path");
+                }
+            }
+            $self->debug_message("Now removing the per lane bam");
```

Comment:
Although the transaction does commit, the merged result is not saved to the database until the "process" context commits.  In the meantime, it might be bad to delete the per-lane BAM.